pinot-plugins/pinot-input-format/pinot-protobuf/src/main/java/org/apache/pinot/plugin/inputformat/protobuf/ProtoBufSchemaUtils.java

```diff
@@ -0,0 +1,180 @@
+      @Nullable TimeUnit timeUnit) {
+    Descriptors.FieldDescriptor.Type fieldType = fieldSchema.getType();
+    if (fieldSchema.isRepeated()) {
+      if (isPrimitiveType(fieldType)) {
+        addFieldToPinotSchema(pinotSchema, valueOf(fieldType), path, false, fieldTypeMap, timeUnit);
+      } else if (fieldsToUnnest.contains(path) && !fieldSchema.isMapField()) {
+        for (Descriptors.FieldDescriptor innerField : fieldSchema.getMessageType().getFields()) {
+          extractSchemaWithComplexTypeHandling(innerField, fieldsToUnnest, delimiter,
+              String.join(delimiter, path, innerField.getName()), pinotSchema, fieldTypeMap, timeUnit);
+        }
+      } else if (!fieldSchema.isMapField()) {
+        addFieldToPinotSchema(pinotSchema, FieldSpec.DataType.STRING, path, true, fieldTypeMap, timeUnit);
+      }
+      // Ignores Map type since it's not supported when complex type handling is enabled
+    } else if (fieldType == Descriptors.FieldDescriptor.Type.MESSAGE) {
+      for (Descriptors.FieldDescriptor innerField : fieldSchema.getMessageType().getFields()) {
+        extractSchemaWithComplexTypeHandling(innerField, fieldsToUnnest, delimiter,
+            String.join(delimiter, path, innerField.getName()), pinotSchema, fieldTypeMap, timeUnit);
+      }
+    } else {
+      FieldSpec.DataType dataType = valueOf(fieldType);
+      addFieldToPinotSchema(pinotSchema, dataType, path, true, fieldTypeMap, timeUnit);
+    }
+  }
+
+  public static FieldSpec.DataType valueOf(Descriptors.FieldDescriptor.Type pinotType) {
+    switch (pinotType) {
+      case INT32:
+      case UINT32:
+      case SINT32:
+      case FIXED32:
+      case SFIXED32:
+        return FieldSpec.DataType.INT;
+      case INT64:
+      case UINT64:
+      case FIXED64:
+      case SINT64:
+      case SFIXED64:
+        return FieldSpec.DataType.LONG;
+      case DOUBLE:
+        return FieldSpec.DataType.DOUBLE;
+      case FLOAT:
+        return FieldSpec.DataType.FLOAT;
+      case BOOL:
+        return FieldSpec.DataType.BOOLEAN;
+      case BYTES:
+        return FieldSpec.DataType.BYTES;
+      case STRING:
+      case ENUM:
+        return FieldSpec.DataType.STRING;
+      default:
+        throw new UnsupportedOperationException("Unsupported ProtoBuf type: " + pinotType);
+    }
+  }
+
+  /**
+   * @return if the given avro type is a primitive type.
+   */
+  public static boolean isPrimitiveType(Descriptors.FieldDescriptor.Type pinotType) {
+    switch (pinotType) {
+      case INT32:
+      case INT64:
+      case UINT32:
+      case UINT64:
+      case SINT32:
+      case SINT64:
+      case FIXED64:
+      case FIXED32:
+      case SFIXED64:
+      case DOUBLE:
+      case FLOAT:
+      case BOOL:
+      case BYTES:
+      case STRING:
+      case ENUM:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private static void addFieldToPinotSchema(Schema pinotSchema, FieldSpec.DataType dataType, String name,
```

Comment:
is this code re-used from somewhere?

Reply:
Yes. This is from AvroUtils and JsonUtils. They both have this function:
[AvroUtils](https://github.com/apache/pinot/blob/a2e85ebf85aba4ced882db35b356e4b18ee37c49/pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroUtils.java#L352)
[JsonUtils](https://github.com/apache/pinot/blob/a2e85ebf85aba4ced882db35b356e4b18ee37c49/pinot-spi/src/main/java/org/apache/pinot/spi/utils/JsonUtils.java#L698)

Reply:
We should add it to a common package and use it. I can do that in a follow up pr.